### ISSUE TITLE
Typescript support

### ISSUE THREAD

**Issue by saidelike** [vaguue/Ghidra.js#1]
This is a cool project. Thank you for writing it.

i was wondering if you have considered providing examples with typescript. It would ease importing other typescript libraries into our code.

**Comment by vaguue** [comment]
Hello! I definitely wanted to do this, but decided to postpone TypeScript support for later because I rarely use it myself. But now that you've suggested it, I'll try to add it soon!

**Comment by saidelike** [comment]
Hi, anything I can do to help with that? That would be really cool and I already have a project that is written in typescript that I could use to test it!

**Comment by vaguue** [comment]
Hi! Sorry for the delay, been busy with my another project [over-the-wire](https://github.com/vaguue/over-the-wire). Well, first of all, we should convert the Ghidra Java documentation (starting with the [currentProgram](https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/Program.html) definition) to a `*.d.ts` file, so type checking will work, secondly we should integrate a typescript building mechanism in the Ghidra itself, I think. Basically, we can check if a file has a `.ts` extension and run something like `tsc` with it and then proceed as with the usual JS scripting routine, so users will be able to run Typescript code from the Ghidra GUI. I can handle the last step easily, so if you will manage to do the first, it would be a great help! I will start integrating typescript compiler in the Ghidra this weekend.

**Comment by saidelike** [comment]
Hi, I've started adding typescript definitions for Program and dependencies in https://github.com/saidelike/Ghidra.js/tree/typescript/typings (e.g. for the [Program](https://github.com/saidelike/Ghidra.js/blob/typescript/typings/Program.d.ts#L40)). Hopefully this is helpful. Feel free to take them and integrate them to your repo or I can do a PR if you prefer.

**Comment by vaguue** [comment]
Thank you, PR would be nice ❤️

**Comment by saidelike** [comment]
Initial PR is https://github.com/vaguue/Ghidra.js/pull/2. What would be the next steps for this?

**Comment by vaguue** [comment]
I'm actually contemplating about this a lot :) Initially I thought it would be a nice idea to add TypeScript support inside the plugin, but on the other hand it will add size to the plugin as well and make building process a little harder. Maybe it's best to delegate the TypeScript building process to the user? Anyway, Your PR will me merged for sure, I will add some more types today and let's go

**Comment by vaguue** [comment]
More about initial plan: I can bundle TypeScript binary inside the plugin, so when you click on your script in the Ghidra GUI it will do all the building things by itself, which is nice.

**Comment by vaguue** [comment]
I thought about it and decided that we should first create a separate package for our module `ghidra.js` so that it can be added as a dependency in TypeScript projects as `@types/ghidra.js`. The structure will look like this:

```typescript
export interface CurrentProgram {
  // define the interface for currentProgram
}

export interface CurrentApi {
  // define the interface for currentApi
}

declare global {
  var currentProgram: CurrentProgram;
  var currentApi: CurrentApi;
}
```

I'll create this package and submit a pull request. For now, I think we should leave the build process on the user's side. I'll prepare an example in the examples folder once the pull request is approved in DefinitelyTyped.

**Comment by saidelike** [comment]
Ya I agree that seems the shortest path to get things working to leave the typescript build on the user's side. We can improve that later imho.

Good idea about adding a separate package for all the types and global. Seems clean way of doing it.

**Comment by vaguue** [comment]
[Made a PR to DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70167), and I invite you to review it.
But our typing are very incomplete, we should think about it. 
I've created a simple script for parsing Java docs: https://github.com/vaguue/Ghidra.js/blob/main/scripts/parseDocs.js, which could be enhanced to parse the whole docs. But first, I think it's a good idea to create typing for built-in Java types.
Once I manage to merge the PR we can test it and than improve the typings.

**Comment by vaguue** [comment]
Btw if you would like to discuss some more stuff about our projects, you can join my newly created discord server: https://discord.gg/KythtrMx

**Comment by vaguue** [comment]
Finally, https://www.npmjs.com/package/@types/ghidra.js

**Comment by vaguue** [comment]
Managed to publish some types

**Comment by saidelike** [comment]
> Btw if you would like to discuss some more stuff about our projects, you can join my newly created discord server: https://discord.gg/KythtrMx

Seems the link has expired. Can you send a new one please?

**Comment by vaguue** [comment]
https://discord.gg/6c52xRdu here it is
